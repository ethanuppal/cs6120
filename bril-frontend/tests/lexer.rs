use bril_frontend::{lexer::Token, loc::Loc};
use insta::assert_debug_snapshot;
use logos::Logos;

#[allow(dead_code)] // since Debug
#[derive(Debug)]
struct LexerSnapshot<'a> {
    code: &'a str,
    tokens: Vec<Loc<Token<'a>>>,
}

#[allow(dead_code)] // since Debug
#[derive(Debug)]
struct LexerErrorSnapshot<'a> {
    code: &'a str,
    tokens: Vec<Loc<Token<'a>>>,
    failure: Loc<&'a str>,
    leftover: &'a str,
}

macro_rules! lexer_snapshot {
    ($name:ident, $code:expr) => {
        #[test]
        fn $name() {
            let code = $code;
            let mut lexer = Token::lexer(code);
            let mut tokens = vec![];
            while let Some(next) = lexer.next() {
                if let Ok(token) = next {
                    tokens.push(Loc::new(token, lexer.span()));
                } else {
                    panic!("Failed to lex. Leftover: {}", lexer.remainder());
                }
            }

            assert_debug_snapshot!(LexerSnapshot { code, tokens });
        }
    };
}

macro_rules! lexer_error {
    ($name:ident, $code:expr) => {
        #[test]
        fn $name() {
            let code = $code;
            let mut lexer = Token::lexer(code);
            let mut tokens = vec![];
            while let Some(next) = lexer.next() {
                if let Ok(token) = next {
                    tokens.push(Loc::new(token, lexer.span()));
                } else {
                    break;
                }
            }

            if lexer.remainder().is_empty() {
                panic!("Lexing was unfortunately successful: {:?}", tokens);
            }

            assert_debug_snapshot!(LexerErrorSnapshot {
                code,
                tokens,
                failure: Loc::new(lexer.slice(), lexer.span()),
                leftover: lexer.remainder()
            });
        }
    };
}

lexer_snapshot! {
    all_tokens_work,
    r#"import from as @main .foo foo "path/to/lol" {}:<>;=,() 5 5.0 'a'"#
}

lexer_snapshot! {
    add_bril_lexes,
    include_str!("../bril-programs/add.bril")
}

lexer_snapshot! {
    import_bril_lexes,
    include_str!("../bril-programs/import.bril")
}

lexer_error! {
    invalid_characters,
    "$main"
}
