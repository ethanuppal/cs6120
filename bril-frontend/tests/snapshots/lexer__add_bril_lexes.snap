---
source: bril-frontend/tests/lexer.rs
expression: "LexerSnapshot { code, tokens }"
---
LexerSnapshot {
    code: "@main {\n  v0: int = const 1;\n  v1: int = const 2;\n  v2: int = add v0 v1;\n  print v2;\n}\n",
    tokens: [
        Loc {
            inner: FunctionName(
                "@main",
            ),
            span: 0..5,
        },
        Loc {
            inner: LeftBrace,
            span: 6..7,
        },
        Loc {
            inner: Identifier(
                "v0",
            ),
            span: 10..12,
        },
        Loc {
            inner: Colon,
            span: 12..13,
        },
        Loc {
            inner: Identifier(
                "int",
            ),
            span: 14..17,
        },
        Loc {
            inner: Equals,
            span: 18..19,
        },
        Loc {
            inner: Identifier(
                "const",
            ),
            span: 20..25,
        },
        Loc {
            inner: Integer(
                1,
            ),
            span: 26..27,
        },
        Loc {
            inner: Semi,
            span: 27..28,
        },
        Loc {
            inner: Identifier(
                "v1",
            ),
            span: 31..33,
        },
        Loc {
            inner: Colon,
            span: 33..34,
        },
        Loc {
            inner: Identifier(
                "int",
            ),
            span: 35..38,
        },
        Loc {
            inner: Equals,
            span: 39..40,
        },
        Loc {
            inner: Identifier(
                "const",
            ),
            span: 41..46,
        },
        Loc {
            inner: Integer(
                2,
            ),
            span: 47..48,
        },
        Loc {
            inner: Semi,
            span: 48..49,
        },
        Loc {
            inner: Identifier(
                "v2",
            ),
            span: 52..54,
        },
        Loc {
            inner: Colon,
            span: 54..55,
        },
        Loc {
            inner: Identifier(
                "int",
            ),
            span: 56..59,
        },
        Loc {
            inner: Equals,
            span: 60..61,
        },
        Loc {
            inner: Identifier(
                "add",
            ),
            span: 62..65,
        },
        Loc {
            inner: Identifier(
                "v0",
            ),
            span: 66..68,
        },
        Loc {
            inner: Identifier(
                "v1",
            ),
            span: 69..71,
        },
        Loc {
            inner: Semi,
            span: 71..72,
        },
        Loc {
            inner: Identifier(
                "print",
            ),
            span: 75..80,
        },
        Loc {
            inner: Identifier(
                "v2",
            ),
            span: 81..83,
        },
        Loc {
            inner: Semi,
            span: 83..84,
        },
        Loc {
            inner: RightBrace,
            span: 85..86,
        },
    ],
}
